integrations/access/accesslist/app.go

```diff
@@ -0,0 +1,251 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package accesslist
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	alclient "github.com/gravitational/teleport/api/client/accesslist"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/accesslist"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/integrations/access/common"
+	"github.com/gravitational/teleport/integrations/access/common/recipient"
+	"github.com/gravitational/teleport/integrations/access/common/teleport"
+	"github.com/gravitational/teleport/integrations/lib"
+	"github.com/gravitational/teleport/integrations/lib/logger"
+	pd "github.com/gravitational/teleport/integrations/lib/plugindata"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+const (
+	// oneWeek is the number of hours in a week.
+	oneWeek = 24 * time.Hour * 7
+)
+
+type App[M MessagingBot] struct {
+	pluginName  string
+	apiClient   teleport.Client
+	accessLists services.AccessListsGetter
+	pluginData  *pd.CompareAndSwap[pd.AccessListNotificationData]
+	bot         M
+	job         lib.ServiceJob
+	clock       clockwork.Clock
+}
+
+func NewApp[M MessagingBot]() *App[M] {
+	app := &App[M]{}
+	app.job = lib.NewServiceJob(app.run)
+	return app
+}
+
+func (a *App[M]) Init(baseApp *common.BaseApp[M]) error {
+	a.pluginName = baseApp.PluginName
+	a.apiClient = baseApp.APIClient
+	a.accessLists = accessListClient(a.apiClient)
+	if a.accessLists == nil {
+		return trace.BadParameter("api client does not contain an access list client")
+	}
+	a.pluginData = pd.NewCAS(
+		a.apiClient,
+		a.pluginName,
+		types.KindAccessList,
+		pd.EncodeAccessListNotificationData,
+		pd.DecodeAccessListNotificationData,
+	)
+	a.bot = baseApp.Bot
+
+	if a.clock == nil {
+		a.clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+func (a *App[_]) Start(ctx context.Context, process *lib.Process) (err error) {
+	process.SpawnCriticalJob(a.job)
+
+	return nil
+}
+
+func (a *App[_]) WaitReady(ctx context.Context) (bool, error) {
+	return a.job.WaitReady(ctx)
+}
+
+func (a *App[_]) WaitForDone() {
+	<-a.job.Done()
+}
+
+func (a *App[_]) Err() error {
+	return nil
+}
+
+// run will monitor access lists and post review reminders.
+func (a *App[_]) run(ctx context.Context) error {
+	process := lib.MustGetProcess(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+
+	// On process termination, explicitly cancel the context
+	process.OnTerminate(func(ctx context.Context) error {
```

Comment:
Can we avoid creating this separate context and registering this termination func and instead just add a case to the select in the loop below that checks for `<-process.Done()`?

Reply:
Currently `<-process.Done()` isn't emitted until `OnTerminate` is called. If we add this into the select loop below, it looks like the plugin sits indefinitely when trying to terminate it. Without refactoring the `process`, which I'm hesitant to do as part of this PR, I don't think I can change this easily.

Reply:
I see. Agree that changing `process` is not in scope here.